Feat: connect explorer settings page

karliatto opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
* New page in connect-explorer to set connect options. For now only `trustedHost` but we can extend it for transport for example. This page is for now intended to be used in e2e tests.
* Extend permissions test to make sure the new feature works.

karliatto commented on packages/connect-web/src/connectSettings.ts:
```diff
@@ -71,6 +71,11 @@ export const parseConnectSettings = (input: Partial<ConnectSettings> = {}): Conn
             settings.trustedHost = false;
         }
+
+        // Even if connect is running on not trusted domain we can override it with `trust-issues=false` query param.
+        if (query['trust-issues'] === 'false') {
+            settings.trustedHost = true;
+        }
```
This is required if we want to support both `trust-issues` as well as setting page in connect-explorer.

In order to guarantee the current tests pass `trust-issues` is top in the hierarchy.

We can change it as we want later on and maybe use only settings page in connect-explorere and remove `trust-issues` from query param.

mroz22 replied:
does this work? it is not really supposed to work https://github.com/trezor/trezor-suite/blob/develop/packages/connect/src/data/connectSettings.ts#L72

dahaca commented on packages/components/src/components/form/Checkbox/Checkbox.tsx:
```diff
@@ -87,13 +89,19 @@ export const Checkbox = ({
 
     return (
         <Wrapper
+            data-test={dataTest}
```
It already accepts `data-test` as a part of `rest`. It's passed [here](https://github.com/trezor/trezor-suite/blob/45f7b8bc1266df82223abfbee20f3f5b61e538eb/packages/suite/src/components/firmware/CheckSeedStep.tsx#L148) for example. :)

karliatto replied:
Cool. I removed that 9da580e1edb8550f24d0d4bcedf138449ef1f283

dahaca commented on packages/components/src/components/form/Checkbox/Checkbox.tsx:
```diff
@@ -87,13 +89,19 @@ export const Checkbox = ({
         >
-            <IconWrapper $color={color} isChecked={isChecked} isDisabled={isDisabled}>
+            <IconWrapper
+                data-test="@checkbox/icon-wrapper"
```
Why is it needed here?

karliatto replied:
Removed, 9da580e1edb8550f24d0d4bcedf138449ef1f283 
It was a leftover from something I was experimenting.

mroz22 commented on packages/connect-popup/e2e/tests/permissions.test.ts:
```diff
@@ -18,23 +19,49 @@ const fixtures = [
+            await waitAndClick(page, ['@checkbox/trustedHost']);
+            await waitAndClick(page, ['@submit-button']);
+        }
         await page.goto(`${url}${f.queryString}#/method/verifyMessage`);
```
page.goto does not cause page to refresh unloading all scripts? just asking..

karliatto replied:
The refresh only initialized connect if it has not been initialized yet, in this case it does not affect because it was initialized. 

You can test it manually or look at the added tests in this PR https://github.com/trezor/trezor-suite/pull/9864/commits/e20b1780eccb0ccfc8b5ae6decd20916607c754f

mroz22 replied:
so this is not reload, only navigation I guess. triggering actions bases on route changes from some middleware, right?

karliatto replied:
Yes, that is right.

mroz22 commented on packages/connect-explorer/src/components/Settings.tsx:
```diff
@@ -0,0 +1,49 @@
+export const Settings = () => {
+    const connectOptions = useSelector(state => ({
+        trustedHost: state.connect?.options?.trustedHost || false,
+        transports: state.connect?.options?.transports || 'bridge',
```
'bridge' does not really match what is allowed by types. maybe types are not strict enough here

karliatto replied:
Yes, let's be simple and use only `trustedHost` now, then we can extend it 336857ebe8d864e45497fedd054b8bd65c0641b3

mroz22 commented on packages/connect-explorer/src/components/Settings.tsx:
```diff
@@ -0,0 +1,49 @@
+
+export const Settings = () => {
+    const connectOptions = useSelector(state => ({
+        trustedHost: state.connect?.options?.trustedHost || false,
```
in general, trusted host should only be "downgradable", so this line feels correct. I am just not sure whether this option is taken into account by connectSettings.ts. is it?

karliatto replied:
Yes it is. Done by the action `onConnectOptionChange` which is passed as `onFieldChange` to the field.

I wanted to reuse the existing mechanisms and it works.

If there is more time to improve connect-explorer code I would like to refactor it a bit the `GetField`  but for now it works.

mroz22 commented on packages/connect-explorer/src/reducers/trezorConnectReducer.ts:
```diff
@@ -46,6 +46,22 @@ const removeDevice = (state: ConnectState, device: TrezorConnectDevice): Connect
     return state;
 };
 
+const onOptionChange = (state: ConnectState, field: Field<any>, value: any): ConnectState => {
```
low priority: maybe value could be typed better?  and maybe field as well 🤔

karliatto replied:
1e33ca1a0731c1c9591385bf1c913359a67e03cb

mroz22 commented on packages/connect-popup/e2e/tests/permissions.test.ts:
```diff
@@ -18,23 +19,49 @@ const fixtures = [
     {
         description: `iframe and host same origins but with trust-issues=true -> show permissions`,
         queryString: '?trust-issues=true',
+        setTrustedHost: false,
```
if we would ensure possibility to downgrade trust level in TrezorConnect.init parameter, do we still have a use case for trust-issues query string param? 

on the other hand - we can pass connectSrc both in init and query string 🤔

karliatto replied:
Yes, that is a good question I also have.
Should we have both ways 
1. set by query param
2. set by connect-explorer new setting page
:question: 

And I think if the reason for using query param `trust-issues` is only for testing I would get rid of that and use only the connect-explorer settings page.

mroz22 replied:
1. trust-issues is only for testing
2. if we can have TrezorConnect.init({ trustedHost: false }) (and false is the only accepted option), then it could do the same. 
3. query string is little bit more convenient, on the other hand, it is not possible to use it in webextensions anyway

So yes, I am slightly in favor of removing trust-issues and implementing its functionality in trezorconnect.init

mroz22 commented on packages/connect-popup/e2e/tests/permissions.test.ts:
```diff
@@ -18,23 +19,49 @@ const fixtures = [
+        description:
+            'iframe and host same origins but with settings to trustedHost -> no permissions shown',
+        queryString: '',
+        setTrustedHost: true,
+        trustedHost: true,
```
there is one extra line? `setTrustedHost` and `trustedHost`, probably some leftover

karliatto replied:
yes, not needed 4b6ffdeaf003b86db13a6fd0c5a85607be70e6f9

mroz22 commented on packages/connect-popup/e2e/tests/permissions.test.ts:
```diff
@@ -18,23 +19,49 @@ const fixtures = [
     },
+    {
+        description:
+            'iframe and host same origins but with settings to trustedHost -> no permissions shown',
```
I might seem retarded, but please once again - we need this because we can employ query string for webextension right? 

just summarizing my thoughts: 
- normally, sldev, localhost, etc are trusted by default. 
- we may downgrade their trust level (useful for tests) either by query string param, or now, newly by init param.
- webextension, on the other hand, does not run on a trusted domain. so, by default it displays permissions screen etc. here you are introducing an option to increase trust level? my question is (and maybe it was already answered elsewhere): how do we guarantee that this is not abused by a 3rd party?

karliatto replied:
Yes, we cannot use query string in web extension so we need other way to simulate the not trusted environment even iframe and popup are running in same domain during tests `http://localhost:8088`

Your points are almost right, but the last one. webextension is running on TRUSTED domain since iframe and popup are running under same domain, but we want to test the scenario where it is not trusted domain and for that we need to do it throw TrezorConnect.init becuase the query param does not work in webextension.

I think it is a mess to have 2 ways of doing the same so I would use only one, the settings page, and maybe we can run all the tests with 2 flows (trusted and not trusted) so we test both flows for every test because now for some tests we are testing with trusted host and other test we are using not trusted hosts.

mroz22 replied:
> webextension is running on TRUSTED domain since iframe and popup are running under same domain, 

Iframe and popup are always running on the same domain. so this point is not valid. I still don't understand how webextension is percieved as trusted here? it must be something specific to testing setup?

mroz22 replied:
its about host vs iframe domain. and since extension is never sldev nor trezor.io (or localhost) it imho can never be trusted.

karliatto replied:
Sorry I mistook it , I wanted to say host instead of popup.

But I found the reason, it is because iframe is running on localhost during the tests then it is consider trusted even the extension is not whitelisted.

https://github.com/trezor/trezor-suite/blob/develop/packages/connect-iframe/src/connectSettings.ts#L44

mroz22 commented on packages/connect-explorer/src/actions/trezorConnectActions.ts:
```diff
@@ -19,6 +23,14 @@ export function onSelectDevice(path: string) {
+        value,
+    },
+});
+
```
yes, this is something we want in general. in the future we would like to have some UI presenting information about options for TrezorConnect.init. So far, lets use it for our need in tests only (incomplete and hidden), once we are finished, we can show it publicly.

karliatto replied:
Yes, I agree we can add more UI sugar :candy:  once we add more options and probably documentation and so on.

karliatto commented:
@mroz22 As we agreed I removed the `trust-issues` and updated the tests that had that.

Now connect-explorer is by default initialized as `trustedHost: false` and we use the the new settings page to make it `trustedHost:true` that we actually only need to use in packages/connect-popup/e2e/tests/permissions.test.ts.

mroz22 reviewed (approved):
yep, now it makes very good sense to me. more over, we have something we are going to incorporate into visible part of connect explorer in the future. some interactive walkthrough "how to" work with connect.
